src/NServiceBus.NHibernate/SynchronizedStorage/NHibernateAmbientTransactionSynchronizedStorageSession.cs

```diff
@@ -9,28 +10,40 @@
     [SkipWeaving]
     class NHibernateAmbientTransactionSynchronizedStorageSession : CompletableSynchronizedStorageSession, INHibernateSynchronizedStorageSession
     {
-        readonly IDbConnection connection;
-        readonly bool ownsConnection;
+        Func<IDbConnection> connectionFactory;
+        Func<IDbConnection, ISession> sessionFactory;
+        ISession session;
+        IDbConnection connection;
 
-        public NHibernateAmbientTransactionSynchronizedStorageSession(ISession session, IDbConnection connection, bool ownsConnection)
+        public NHibernateAmbientTransactionSynchronizedStorageSession(Func<IDbConnection> connectionFactory, Func<IDbConnection, ISession> sessionFactory)
         {
-            this.connection = connection;
-            this.ownsConnection = ownsConnection;
-            Session = session;
+            this.connectionFactory = connectionFactory;
+            this.sessionFactory = sessionFactory;
         }
 
-        public ISession Session { get; }
-        public void Dispose()
+        public ISession Session
         {
-            Session.Flush();
-            Session.Dispose();
-
-            if (ownsConnection)
+            get
             {
-                connection.Dispose();
+                if (connection == null)
```

Comment:
In theory creating a connection and opening a session are two operations and the failure might happen between these calls. I guess in normal scenario the flow would be interrupted anyway but if somebody does:

```
while (!success)
{
try
{
   var s = session.Session;
   success = true;
}
catch (Exception) {}
}
```

then this code will behave better. Hmm.. after writing this I think that would be a very unusual scenario. I'll collapse these ifs.